### PR TITLE
Removed recursion in place for iterable solution to prevent stack overflow

### LINKED ISSUE #2387: [Bug]: RangeError: Maximum call stack size exceeded. when openning the application
### Verified issue does not already exist?

- [X] I have searched and found no existing issue

### Is this related to GoCardless, Simplefin or another bank-sync provider?

- [ ] I have checked my server logs and could not see any errors there
- [ ] I will be attaching my server logs to this issue
- [ ] I will be attaching my client-side (browser) logs to this issue
- [ ] I understand that this issue will be automatically closed if insufficient information is provided

### What happened?

 Imported some OFX with no problems, then I imported an CSV, categorized its transactions. Tried to go to budget, this error starts happening. Now I'm unable to open the application at all.

### What error did you receive?

graph-data-structure.ts:79 Uncaught (in promise) RangeError: Maximum call stack size exceeded
at ep (graph-data-structure.ts:79:13)
at ep (graph-data-structure.ts:85:9)
at ep (graph-data-structure.ts:85:9)
at ep (graph-data-structure.ts:85:9)
at ep (graph-data-structure.ts:85:9)
at ep (graph-data-structure.ts:85:9)
at ep (graph-data-structure.ts:85:9)
at ep (graph-data-structure.ts:85:9)
at ep (graph-data-structure.ts:85:9)
at ep (graph-data-structure.ts:85:9)

### Where are you hosting Actual?

Docker

### What browsers are you seeing the problem on?

Chrome, Microsoft Edge

### Operating System

Windows 11

### PR DESCRIPTION
**Pull request by lelemm**
Hi, I'm trying to solve the issue #2387 with this fix. Can you check this @MatissJanis?
The idea is not depend on recursion to sort.

fix #2387

### CODE REVIEW

**Comment by lelemm** [comment]
Noticed the error in tests. Will investigate

**Comment by MatissJanis** [comment]
This is a pretty risky change. How could we minimize the risk?

Here's an idea:

What if we kept both algorithms and ran them in parellel. And then compared the end-result. If the result is different: show an error message (so folks can report it via a github issue) + use the recursive algorithm.

That would give us higher confidence in the release.

Another option would be to put it under a feature flag. And allow people to opt-out of the iterative solution if it breaks something. And then eventually sunset the feature flag is no issues reported.

WDYT?

**Comment by lelemm** [comment]
> This is a pretty risky change

Agreed. The idea to put under a feature flag is nice.

Or we can try to detect when the stack overflow is about happen (set a stack limit for example), abort the execution of the recursive version and go for iterable version

**Comment by lelemm** [comment]
Ultimately, the main purpose of this change is prevent the application to stop opening because of a stack overflow.

**Comment by lelemm** [comment]
I'm changing back to WIP so we can discuss the best way to move forward

**Comment by MatissJanis** [comment]
Long-term we should have only 1x logic here instead of 2x. So I would not recommend the approach of detecting runtime exceptions.

**Comment by lelemm** [comment]
Right. So we could go with a feature flag for the iterative version.

And for safe guard, put could put a limit of possible recursions. If this limit is reached, the spreadsheet generation hangs, a error message shows in actual telling the user that there is an error with their transactions.
This is ultimately to prevent an error that keeps the user from using the application.

What do you think? Can I go with this path?

**Comment by MatissJanis** [comment]
Thanks sounds very good! Thanks for being so considerate! 🙇

**Comment by lelemm** [comment]
@MatissJanis added the feature flag and added recursion limit to 2500

File: packages/desktop-client/src/hooks/useFeatureFlag.ts [hunk 1]
```diff
@@ -9,6 +9,7 @@ const DEFAULT_FEATURE_FLAG_STATE: Record<FeatureFlag, boolean> = {
   customReports: false,
   spendingReport: false,
   simpleFinSync: false,
+  iterableTopologicalSort: false,
```
Comment (MatissJanis) [inline packages/desktop-client/src/hooks/useFeatureFlag.ts:12]: **💬 suggestion:** ‏IMO we want your new iterable solution to be enabled as the default. If we receive bug reports: we can point folks to disable this feature flag. And if there are no bug reports: a few releases later we can clean up this feature flag and just keep your new logic. WDYT?
```suggestion
  iterableTopologicalSort: true,
```

Comment (lelemm) [reply packages/desktop-client/src/hooks/useFeatureFlag.ts]: Sure, just changed it

**Review by MatissJanis — APPROVED** [review]
LGTM! Tested with my real (large) budget file and also with the `crash.csv` file that was previously not working.

Is there anything else you'd like to do here? The PR still has "[WIP]" prefix.

**Comment by lelemm** [comment]
I forgot to remove WIP. ops :)

**Comment by MatissJanis** [comment]
Thanks for your contribution! Really looking forward to removing the recursive solution a few releases down the line :)

**Comment by metabubble** [comment]
Was it intended that the feature flag has been activated automatically? After every update, I check what new experimental features they are and I noticed this was auto-activated.

**Pull request merged**